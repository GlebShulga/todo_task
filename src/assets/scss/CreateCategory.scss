.CreateCategory {
  // display: grid;
  justify-items: left;
  padding: 1rem 1rem;
}

.CreateCategory-form {
  display: flex;
  overflow: hidden;
}

.CreateCategory-form_input {
  padding: 0.5rem 1rem;
  border: lightgrey;
  background-color: lightgrey;
  outline: none;

  &:focus {
    background-color: darkgrey;
  }
}

.CreateCategory-form_button_add {
  padding: 0.5rem 1rem;
  background-color: limegreen;
  color: white;
  font-weight: 700;
  &:hover {
    background-color: lightgreen;
  }
  }

.CreateCategory-form_button_close {
  padding: 0.5rem 1rem;
  margin: 0 0.1rem;
  background-color: grey;
  color: white;
  font-weight: 700;
  &:hover {
    background-color: lightslategray;
  }
}

.CreateCategory-form_button_error {
  opacity: 0.5;
}

.Error {
  color: red;
  padding-top: 1rem;
  text-align: center;
}
