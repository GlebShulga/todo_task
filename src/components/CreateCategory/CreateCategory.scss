.create_category {
  justify-items: left;

}

.create_category-form {
  display: flex;
  overflow: hidden;
}

.create_category-form_input {
  padding: 0.5rem 1rem;
  border: lightgrey;
  background-color: lightgrey;
  outline: none;

  &:focus {
    background-color: darkgrey;
  }
}

.form_button_add {
  padding: 0.5rem 1rem;
  background-color: limegreen;
  color: white;
  font-weight: 700;
  &:hover {
    background-color: lightgreen;
  }
}

.create_category-form_button_close {
  padding: 0.5rem 1rem;
  margin: 0 0.1rem;
  background-color: grey;
  color: white;
  font-weight: 700;
  &:hover {
    background-color: lightslategray;
  }
}

.error {
  color: red;
  padding-top: 1rem;
  text-align: center;
  margin: auto;
}
