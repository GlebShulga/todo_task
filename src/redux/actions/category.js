/* eslint-disable no-loop-func */
import axios from "axios";
import {
  ADD_CATEGORY,
  GET_CATEGORIES_LIST,
  GET_CATEGORIES_TITLE_LIST,
  GET_ROOT_CATEGORY_LIST,
  RENAME_CATEGORY,
  DEL_CATEGORY,
  UPDATE_CHOSEN_CATEGORY,
  SET_IS_FILTER_STATUS_DONE,
  SET_IS_CREATE_TASK_MODAL,
} from "../types/category";

export function fetchCategoryList() {
  return (dispatch, getState) => {
    const store = getState();
    const url = store.router.location.pathname?.substring(1);
    const complexUrl = url?.match("[^/](.*?)(?=/)");
    const categoryUrl = complexUrl !== null ? complexUrl[0] : url;
    return axios
      .get("/api/v1/category")
      .then(({ data }) => {
        let current = data.find((cat) => cat.categoryTitle === categoryUrl);
        let updatedCategoryList = data;
        while (current?.parentCategoryId) {
          updatedCategoryList = updatedCategoryList.map((cat) =>
            cat.categoryId === current.parentCategoryId
              ? { ...cat, isExpanded: true }
              : cat
          );
          current = updatedCategoryList.find(
            (cat) => cat.categoryId === current.parentCategoryId
          );
        }

        const rootList = updatedCategoryList.filter(
          (el) => el.parentCategoryId === null
        );
        const titleList = data.map((category) => category.categoryTitle);
        dispatch({ type: GET_ROOT_CATEGORY_LIST, rootList });
        dispatch({
          type: GET_CATEGORIES_LIST,
          categoryList: updatedCategoryList,
        });
        dispatch({ type: GET_CATEGORIES_TITLE_LIST, titleList });
      })
      .catch((err) => console.log(err));
  };
}

export function addCategory(categoryTitle, parentCategoryId, lvl) {
  return (dispatch) => {
    axios({
      method: "post",
      url: "/api/v1/category/:categoryTitle",
      data: {
        categoryTitle,
        parentCategoryId,
        lvl,
      },
    })
      .then(({ data: categoryList }) => {
        const rootList = categoryList?.filter(
          (el) => el.parentCategoryId === null
        );
        dispatch({ type: ADD_CATEGORY, categoryList });
        dispatch({ type: GET_ROOT_CATEGORY_LIST, rootList });
      })
      .catch((err) => console.log(err));
  };
}

export function renameCategory(categoryId, categoryTitle) {
  return (dispatch) => {
    axios({
      method: "patch",
      url: "/api/v1/category",
      data: {
        categoryId,
        categoryTitle,
      },
    })
      .then(({ data: categoryList }) => {
        dispatch({ type: RENAME_CATEGORY, categoryList });
      })
      .catch((err) => console.log(err));
  };
}

export function delCategory(categoryId) {
  return (dispatch) => {
    axios({
      method: "delete",
      url: "/api/v1/category",
      data: {
        categoryId,
      },
    }).then(({ data: categoryList }) => {
      const rootList = categoryList.filter(
        (el) => el.parentCategoryId === null
      );
      dispatch({ type: DEL_CATEGORY, categoryList });
      dispatch({ type: GET_ROOT_CATEGORY_LIST, rootList });
    });
  };
}

export function updateChosenCategory(chosenCategory) {
  return { type: UPDATE_CHOSEN_CATEGORY, chosenCategory };
}

export function setIsFilterStatusDone(data) {
  return { type: SET_IS_FILTER_STATUS_DONE, data };
}

export function setIsCreateTaskModalOpen(data) {
  return { type: SET_IS_CREATE_TASK_MODAL, data };
}
