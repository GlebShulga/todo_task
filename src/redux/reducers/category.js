import {
  ADD_CATEGORY,
  GET_CATEGORIES_LIST,
  GET_CATEGORIES_TITLE_LIST,
  GET_ROOT_CATEGORY_LIST,
  RENAME_CATEGORY,
  DEL_CATEGORY,
  UPDATE_CHOSEN_CATEGORY,
  SET_IS_FILTER_STATUS_DONE,
  SET_IS_CREATE_TASK_MODAL,
  SET_IS_OPEN_TASK_TABLE,
} from "../types/category";

const initialState = {
  categoryList: [],
  rootCategoryList: [],
  categoryTitleList: [],
  chosenCategory: {},
  isFilterStatusDone: false,
  isCreateTaskModalOpen: false,
  isOpenTaskTable: false,
};

// eslint-disable-next-line import/no-anonymous-default-export
export default (state = initialState, action) => {
  switch (action.type) {
    case GET_CATEGORIES_LIST:
    case ADD_CATEGORY:
    case DEL_CATEGORY:
    case RENAME_CATEGORY: {
      return { ...state, categoryList: action.categoryList };
    }
    case GET_ROOT_CATEGORY_LIST: {
      return { ...state, rootCategoryList: action.rootList };
    }
    case GET_CATEGORIES_TITLE_LIST: {
      return { ...state, categoryTitleList: action.titleList };
    }
    case UPDATE_CHOSEN_CATEGORY: {
      return { ...state, chosenCategory: action.chosenCategory };
    }
    case SET_IS_FILTER_STATUS_DONE: {
      return { ...state, isFilterStatusDone: action.data };
    }
    case SET_IS_CREATE_TASK_MODAL: {
      return { ...state, isCreateTaskModalOpen: action.data };
    }
    case SET_IS_OPEN_TASK_TABLE: {
      return { ...state, isOpenTaskTable: action.data };
    }
    default:
      return state;
  }
};


