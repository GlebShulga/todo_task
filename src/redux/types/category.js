const ADD_CATEGORY = "ADD_CATEGORY";
const GET_CATEGORIES_LIST = "GET_CATEGORIES_LIST";
const GET_CATEGORIES_TITLE_LIST = "GET_CATEGORIES_TITLE_LIST";
const GET_ROOT_CATEGORY_LIST = "GET_ROOT_CATEGORY_LIST";
const RENAME_CATEGORY = "RENAME_CATEGORY";
const DEL_CATEGORY = "DEL_CATEGORY";
const UPDATE_CHOSEN_CATEGORY = "UPDATE_CHOSEN_CATEGORY";
const SET_IS_FILTER_STATUS_DONE = "SET_IS_FILTER_STATUS_DONE";
const SET_IS_CREATE_TASK_MODAL = "SET_IS_CREATE_TASK_MODAL";
const SET_IS_OPEN_TASK_TABLE = "SET_IS_OPEN_TASK_TABLE";

module.exports = {
  ADD_CATEGORY,
  GET_CATEGORIES_LIST,
  GET_CATEGORIES_TITLE_LIST,
  GET_ROOT_CATEGORY_LIST,
  RENAME_CATEGORY,
  DEL_CATEGORY,
  UPDATE_CHOSEN_CATEGORY,
  SET_IS_FILTER_STATUS_DONE,
  SET_IS_CREATE_TASK_MODAL,
  SET_IS_OPEN_TASK_TABLE,
};
