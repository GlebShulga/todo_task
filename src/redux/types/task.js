const ADD_TASK = "ADD_TASK";
const GET_TASK_LIST = "GET_TASK_LIST";
const PATCH_TASK = "PATCH_TASK";
const UPDATE_CHOSEN_TASK = "UPDATE_CHOSEN_TASK";
const SET_EDITING_TASK_MODE = "SET_EDITING_TASK_MODE ";
const SET_NEW_CATEGORY_FOR_TASK = "SET_NEW_CATEGORY_FOR_TASK";
const SET_IS_SEARCH = "SET_IS_SEARCH";
const SET_SEARCH_CRITERIA = "SET_SEARCH_CRITERIA";

module.exports = {
  ADD_TASK,
  GET_TASK_LIST,
  PATCH_TASK,
  UPDATE_CHOSEN_TASK,
  SET_EDITING_TASK_MODE,
  SET_NEW_CATEGORY_FOR_TASK,
  SET_IS_SEARCH,
  SET_SEARCH_CRITERIA,
};
