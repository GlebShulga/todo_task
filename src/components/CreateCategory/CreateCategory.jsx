import React, { useState } from "react";
import { useSelector, useDispatch } from "react-redux";
import { addCategory } from "../../redux/actions/category";
import "./CreateCategory.scss";
import { setIsCreateTaskModalOpen } from "../../redux/actions/category";

const CreateCategory = () => {
  const dispatch = useDispatch();
  const { chosenCategory, categoryTitleList, isCreateTaskModalOpen } =
    useSelector((s) => s.category);
  const [categoryTitle, setCategoryTitle] = useState("");
  const [lengthError, setLengthError] = useState(false);
  const [categoryAlreadyExist, setCategoryAlreadyExist] = useState(false);

  const onChange = (e) => {
    setCategoryTitle(e.target.value);
    setLengthError(false);
    setCategoryAlreadyExist(false);
  };

  const parentCategoryId = chosenCategory?.categoryId ?? null;

  const lvl = chosenCategory?.lvl === undefined ? 0 : chosenCategory?.lvl + 1;

  const onClickAddCategory = () => {
    if (
      categoryTitle?.trim().length <= 20 &&
      categoryTitle?.trim().length >= 3
    ) {
      if (categoryTitleList.indexOf(categoryTitle) === -1) {
        dispatch(addCategory(categoryTitle, parentCategoryId, lvl));
        setCategoryTitle("");
        if (isCreateTaskModalOpen) {
          dispatch(setIsCreateTaskModalOpen(false));
        }
      } else {
        setCategoryAlreadyExist(true);
      }
    } else {
      setLengthError(true);
    }
  };
  const handleKeypress = (e) => {
    if (e.key === "Enter") {
      onClickAddCategory();
    }
  };

  return (
    <div className="create_category">
      <div className="create_category-form">
        <input
          className="create_category-form_input"
          type="text"
          value={categoryTitle}
          onChange={onChange}
          placeholder="Add new category"
          onKeyPress={handleKeypress}
        />
        {isCreateTaskModalOpen && (
          <button
            type="button"
            className="create_category-form_button_close"
            onClick={() => {
              dispatch(setIsCreateTaskModalOpen(false));
            }}
          >
            Close
          </button>
        )}
        <button
          type="button"
          className="form_button_add"
          onClick={onClickAddCategory}
        >
          Add
        </button>
      </div>
      {lengthError && (
        <div className="error">
          <div> The task length must not be shorter than 3 characters</div>
          <div>and</div>
          <div>must not exceed 20 characters.</div>
        </div>
      )}
      {categoryAlreadyExist && (
        <div className="error">
          <div> Channel`s name already exist</div>
        </div>
      )}
    </div>
  );
};

export default CreateCategory;
