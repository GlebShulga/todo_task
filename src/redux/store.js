import { configureStore } from "@reduxjs/toolkit";
import category from "./reducers/category";
import task from "./reducers/task"

export const store = configureStore({
  reducer: {
    category,
    task
  },
});
