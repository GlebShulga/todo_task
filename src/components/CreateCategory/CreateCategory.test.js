import React from "react";
import { render, fireEvent } from "@testing-library/react";
import axios from "axios";
import store from "../../redux";
import { GET_CATEGORIES_TITLE_LIST } from "../../redux/types/category";
import CreateCategory from "./CreateCategory";
import AppTest from "../../../tests/AppTest";

jest.mock("axios");

describe("<CreateCategory />", () => {
  it("should validate category", () => {
    const { getByTestId, getByText } = render(
      <AppTest>
        <CreateCategory />
      </AppTest>
    );

    fireEvent.click(getByTestId("AddCategoryButton"));
    const error = "The task length must not be shorter than 3 characters";
    expect(getByText(error)).toBeTruthy();
  });

  it("should add new category on click event", () => {
    axios.mockResolvedValue({});

    store.dispatch({
      type: GET_CATEGORIES_TITLE_LIST,
      titleList: ["category_1", "category_2"],
    });

    const { getByTestId } = render(
      <AppTest>
        <CreateCategory />
      </AppTest>
    );

    fireEvent.change(getByTestId("AddCategoryInputField"), {
      target: { value: "Demo category" },
    });
    fireEvent.click(getByTestId("AddCategoryButton"));

    expect(axios).toHaveBeenCalledWith({
      data: {
        categoryTitle: "Demo category",
        parentCategoryId: null,
        lvl: 0,
      },
      method: "post",
      url: "/api/v1/category/:categoryTitle",
    });
  });
});
