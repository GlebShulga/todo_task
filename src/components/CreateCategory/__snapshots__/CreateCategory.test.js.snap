// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CreateCategory Component CreateCategory snapshot 1`] = `
<div
  className="create_category"
>
  <div
    className="create_category-form"
  >
    <input
      className="create_category-form_input"
      onChange={[Function]}
      onKeyPress={[Function]}
      placeholder="Add new category"
      type="text"
      value=""
    />
    <button
      className="form_button_add"
      onClick={[Function]}
      type="button"
    >
      Add
    </button>
  </div>
</div>
`;

exports[`CreateCategory Component CreateCategory snapshot 2`] = `
"Snapshot Diff:
- First value
+ Second value

@@ -18,6 +18,19 @@
        type=\\"button\\"
      >
        Add
      </button>
    </div>
+   <div
+     className=\\"error\\"
+   >
+     <div>
+        The task length must not be shorter than 3 characters
+     </div>
+     <div>
+       and
+     </div>
+     <div>
+       must not exceed 20 characters.
+     </div>
+   </div>
  </div>"
`;
